### PR TITLE
Feature/add option to skip loading claims from profile url

### PR DESCRIPTION
**Pull request by nilsgstrabo**
## Description

Added a new flag `skip-claims-from-profile-url` that controls if ProviderData will provide ClaimExtractor with the ProfileURL (resolved from OIDC metadata or manually set). If set to true, ProfileURL is not provided, and ClaimExtractor will not call the profile URL endpoint to resolve claims missing in the id_token.

The default value is `false`, and should therefore not break default behavior.

## Motivation and Context

When using login.microsoftonline.com as OIDC issuer (e.g. https://login.microsoftonline.com/common/v2.0) and you define a scope (e.g. openid profile offline_access 6dae42f8-4368-4678-94ff-3960e28e3630/user.read) with a custom API resource, the access_token received from the `token` endpoint cannot be used as a bearer token with the userinfo (profile URL) endpoint; The `aud` claim is wrong. When ClaimExtractor calls the userinfo endpoint a 401 is returned, which in turn cause oauth2-proxy to return 500 to the client.

The new flag `skip-claims-from-profile-url` instructs the ProviderData to not pass the ProfileURL to ClaimExtractor, thereby preventing a request to the ProfileUR and a subsequent 401 error. The flag is only required when a custom API resource is defined in the scope.

The change proposed in this PR should solve #1666  and #2314 

## How Has This Been Tested?

As mentioned, when using oauth2-proxy without defining a custom resources in scope, everything works fine, for example:
```
    image: quay.io/oauth2-proxy/oauth2-proxy:v7.5.1
    environment:
      - OAUTH2_PROXY_SCOPE=openid profile offline_access email
      - OAUTH2_PROXY_OIDC_ISSUER_URL=https://login.microsoftonline.com/3aa4a235-b6e2-48d5-9195-7fcf05b459b0/v2.0
      - OAUTH2_PROXY_UPSTREAMS=http://proxy:8080
```

But if I add a custom resource to the scope it fails:
```
  auth:
    image: quay.io/oauth2-proxy/oauth2-proxy:v7.5.1
    environment:
      - OAUTH2_PROXY_SCOPE=openid profile offline_access email 6dae42f8-4368-4678-94ff-3960e28e3630/user.read
      - OAUTH2_PROXY_OIDC_ISSUER_URL=https://login.microsoftonline.com/3aa4a235-b6e2-48d5-9195-7fcf05b459b0/v2.0
      - OAUTH2_PROXY_UPSTREAMS=http://proxy:8080
```
with the following log entries from the oauth2-proxy
```
radix-auth_container        | [2023/11/21 09:20:03] [oauthproxy.go:834] Error redeeming code during OAuth2 callback: could not get claim "groups": failed to fetch claims from profile URL: error making request to profile URL: unexpected status "401": {"error":{"code":"InvalidAuthenticationToken","message":"Access token validation failure. Invalid audience.","innerError":{"date":"2023-11-21T09:20:03","request-id":"1a60f131-bd91-4337-a8d9-59efc4456cfe","client-request-id":"1a60f131-bd91-4337-a8d9-59efc4456cfe"}}}
radix-auth_container        | 172.24.0.1:57348 - ef160e37-444e-454a-be7e-909f2a3abe93 - - [2023/11/21 09:20:02] localhost:8000 GET - "/oauth2/callback?code=<deprectaed>&state=Z8DwjH0hu4HMdVH6p2rPkIeqi1vctezHO28oVBTpDWA%3a%2f&session_state=6c6765ae-b9bc-41bc-a31e-c5e13adb2be8" HTTP/1.1 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36" 500 2771 0.450
```
because the access_token has a `aud` claim not intended for the userinfo endpoint.

If I use my own image built with the new option, everything works as intended (at least for me)
```
  auth:
    image: nilsgustavstrabo/oauth2-proxy:fix
    environment:
      - OAUTH2_PROXY_SCOPE=openid profile offline_access email 6dae42f8-4368-4678-94ff-3960e28e3630/user.read
      - OAUTH2_PROXY_OIDC_ISSUER_URL=https://login.microsoftonline.com/3aa4a235-b6e2-48d5-9195-7fcf05b459b0/v2.0
      - OAUTH2_PROXY_UPSTREAMS=http://proxy:8080
      - OAUTH2_PROXY_SKIP_CLAIMS_FROM_PROFILE_URL=true
```
If I set `OAUTH2_PROXY_SKIP_CLAIMS_FROM_PROFILE_URL=false`, or omit it, it falls back do default behavior and fails the same way as v7.5.1 does.

## Checklist:

Not sure what tests to update. Existing tests in `options` package seems very limited, at least when I look for tests for other options.

I could add tests to ClaimExtractor if I pass the SkipClaimsFromProfileURL as an argument, and let ClaimExtractor decided if it should call the provided profileURL or not. The way is works now is that ProviderData passes nil as profileUrl if `skip-claims-from-profile-url` is true.

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [x] I have written tests for my code changes.

### CODE REVIEW

**Comment by nilsgstrabo** [comment]
Any input/feedback from anyone?

**Comment by nilsgstrabo** [comment]
Any update @kvanzuijlen?
It's ok if you want to reject the PR. No hard feelings :) 
In that case we'll just build from our fork since we are stuck on an old version.

**Comment by kvanzuijlen** [comment]
Hi @nilsgstrabo,

Sorry, I completely overlooked your response. Thanks for the detailed explanation. I'd be curious what other maintainers think as well. @tuunit @JoelSpeed any comments?

**Comment by tuunit** [comment]
Hi, will check it this evening!

File: providers/provider_data.go [hunk 1]
```diff
@@ -283,7 +284,12 @@ func (p *ProviderData) buildSessionFromClaims(rawIDToken, accessToken string) (*
 }
 
 func (p *ProviderData) getClaimExtractor(rawIDToken, accessToken string) (util.ClaimExtractor, error) {
-	extractor, err := util.NewClaimExtractor(context.TODO(), rawIDToken, p.ProfileURL, p.getAuthorizationHeader(accessToken))
+	var profileURL *url.URL
+	if !p.SkipClaimsFromProfileURL {
+		profileURL = p.ProfileURL
+	}
```
Comment (tuunit) [inline providers/provider_data.go:290]: What's your opinion on the following for better readability?

```suggestion
	profileURL := p.ProfileURL
	if p.SkipClaimsFromProfileURL {
		profileURL = &url.URL{}
	}
```

**Review by tuunit — COMMENTED** [review]
Overall, a really well written PR. Thanks a lot for all the effort you put into your description of the issue.

File: providers/provider_data.go [hunk 2]
```diff
@@ -283,7 +284,12 @@ func (p *ProviderData) buildSessionFromClaims(rawIDToken, accessToken string) (*
 }
 
 func (p *ProviderData) getClaimExtractor(rawIDToken, accessToken string) (util.ClaimExtractor, error) {
```
Comment (tuunit) [inline providers/provider_data.go:286]: Regarding testing. Could you take some time a extend the test case `TestProviderData_buildSessionFromClaims` inside the `provider_data_test.go`

We should be able to set the new skip option inside the tests and get a valid response:
https://github.com/oauth2-proxy/oauth2-proxy/blob/63727103db4a8bb22872e8e82cb5fe38018ffbad/providers/provider_data_test.go#L427-L441

**Review by tuunit — CHANGES_REQUESTED** [review]

File: providers/provider_data.go (hunk 1 above)
Comment (nilsgstrabo) [reply providers/provider_data.go:290]: Agree. Committing your suggestion.

File: providers/provider_data.go (hunk 2 above)
Comment (nilsgstrabo) [reply providers/provider_data.go:286]: Thanks, I will take a look.

Comment (nilsgstrabo) [reply providers/provider_data.go:286]: Testing SkipClaimsFromProfileURL turned out to be a bit trickier than expected. The only feasible way I can see, is to use httptest.Server, set ProfileURL (and AccessToken + getAuthorizationHeaderFunc), and test if the server receives a request based on the value for SkipClaimsFromProfileURL. Ref commit 7dc8c14d51276a206dd8d13dcc20977714bcae42

Please let me know if you have any better ideas or suggestions.

File: providers/provider_data_test.go [hunk 3]
```diff
@@ -408,11 +417,42 @@ func TestProviderData_buildSessionFromClaims(t *testing.T) {
...
 	for testName, tc := range testCases {
 		t.Run(testName, func(t *testing.T) {
 			g := NewWithT(t)
 
+			var (
+				profileURL        *url.URL
+				profileURLTracker mock.Mock
+			)
+			if tc.SetProfileURL {
+				profileURLSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (tuunit) [inline providers/provider_data_test.go:442]: I'm fine with setting up a http test server here. We do something quite similar in `claim_extractor_test.go` for the profileURL as well.

Comment (tuunit) [reply providers/provider_data_test.go:442]: But do we really need to introduce a new go dependency for this "simple" test?

Couldn't we just introduce a simple custom handler like in `claim_extractor_test.go#L556`:

https://github.com/oauth2-proxy/oauth2-proxy/blob/52ad31752d0d0b38673d419e56ddfc4cb2f56857/pkg/providers/util/claim_extractor_test.go#L556-L564


This way we don't need to introduce a new dependency to objx and mock.Mock and we will not need to the assertion in line 485

**Review by tuunit — CHANGES_REQUESTED** [review]
@nilsgstrabo 
Whats your opinion?

File: providers/provider_data_test.go (hunk 3 above)
Comment (nilsgstrabo) [reply providers/provider_data_test.go:438]: Agree. The mock.Mock package is completely overkill here. Only need to test that a request to profileURL is sent or not.
Fixed in ceccfb2141ddc63f9728097a4c4ae1e4f8cefd9a

**Review by tuunit — APPROVED** [review]
LGTM

**Comment by nilsgstrabo** [comment]
@tuunit Is there anything else I need to do with this PR? Will it be included in the next release, and do you know approximately when the next release is planned?

**Review by JoelSpeed — DISMISSED** [review]

**Comment by JoelSpeed** [comment]
Code generation needs to be updated here

**Comment by nilsgstrabo** [comment]
> Code generation needs to be updated here

Anything I need to do?

**Comment by JoelSpeed** [comment]
Please run `make generate` and then commit the changes

**Comment by nilsgstrabo** [comment]
Done and pushed

**Pull request merged**